plugins/state_history_plugin/include/eosio/state_history_plugin/session.hpp

```diff
@@ -53,6 +55,11 @@ class session final : public session_base {
       awake_if_idle();
    }
 
+   // allow main thread to drain the strand before destruction -- some awake_if_idle() post()s may be inflight
+   void drain_strand() {
+      boost::asio::post(strand, boost::asio::use_future([](){})).get();
+   }
```

Comment:
Can be `post_async_task()` on 1.2

Reply:
Also,
It might look like we can just do this in the dtor and avoid a special call (which is certainly ugly), but we can't: during nodeos shutdown the thread pools will be stopped before destruction takes place. In that case, trying to drain the strand will never complete since the strand is stopped.